File: src/test/java/com/joecollins/graphics/components/MultiSummaryFrameBuilderTest.java
package com.joecollins.graphics.components;

import static org.junit.Assert.assertEquals;

import com.joecollins.bindings.Bindable;
import com.joecollins.bindings.Binding;
import com.joecollins.models.general.Party;
import java.awt.Color;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import org.apache.commons.lang3.tuple.ImmutablePair;
import org.apache.commons.lang3.tuple.Pair;
import org.junit.Test;

public class MultiSummaryFrameBuilderTest {

  @Test
  public void testTooCloseToCallBuilder() {
    Party yp = new Party("Yukon Party", "YP", Color.BLUE);
    Party ndp = new Party("New Democratic Party", "NDP", Color.ORANGE);
    Party lib = new Party("Liberal", "LIB", Color.RED);
    Party grn = new Party("Green", "GRN", Color.GREEN);
    Party ind = new Party("Independent", "IND", Color.GRAY);

    List<Riding> ridings = new ArrayList<>();
    ridings.add(new Riding("Klondike")); // 0
    ridings.add(new Riding("Kluane")); // 1
    ridings.add(new Riding("Lake Laberge")); // 2
    ridings.add(new Riding("Mayo-Tatchun")); // 3
    ridings.add(new Riding("Mount Lorne-Southern Lakes")); // 4
    ridings.add(new Riding("Pelly-Nisutlin")); // 5
    ridings.add(new Riding("Vuntut Gwitchin")); // 6
    ridings.add(new Riding("Watson Lake")); // 7
    MultiSummaryFrame frame =
        MultiSummaryFrameBuilder.tooClose(
                ridings,
                Riding::isTooClose,
                Riding::getMargin,
                r -> Binding.fixedBinding(r.name.toUpperCase()),
                Riding::getBoxes,
                2)
            .withHeader(() -> "TOO CLOSE TO CALL")
            .build();
    assertEquals("TOO CLOSE TO CALL", frame.getHeader());
    assertEquals(0, frame.getNumRows());

    // add first (36)
    ridings.get(5).setResults(Map.of(yp, 140, ndp, 104, lib, 76, grn, 11));
    assertEquals(1, frame.getNumRows());
    assertEquals("PELLY-NISUTLIN", frame.getRowHeader(0));
    assertEquals(Color.BLUE, frame.getColor(0, 0));
    assertEquals("YP: 140", frame.getValue(0, 0));

    // add to top (3/36)
    ridings.get(6).setResults(Map.of(yp, 35, ndp, 2, lib, 38));
    assertEquals(2, frame.getNumRows());
    assertEquals("VUNTUT GWITCHIN", frame.getRowHeader(0));
    assertEquals(Color.RED, frame.getColor(0, 0));
    assertEquals("LIB: 38", frame.getValue(0, 0));
    assertEquals("PELLY-NISUTLIN", frame.getRowHeader(1));
    assertEquals(Color.BLUE, frame.getColor(1, 0));
    assertEquals("YP: 140", frame.getValue(1, 0));

    // add beyond limit (3/36/40)
    ridings.get(7).setResults(Map.of(yp, 150, ndp, 110, lib, 106, ind, 19));
    assertEquals(2, frame.getNumRows());
    assertEquals("VUNTUT GWITCHIN", frame.getRowHeader(0));
    assertEquals(Color.RED, frame.getColor(0, 0));
    assertEquals("LIB: 38", frame.getValue(0, 0));
    assertEquals("PELLY-NISUTLIN", frame.getRowHeader(1));
    assertEquals(Color.BLUE, frame.getColor(1, 0));
    assertEquals("YP: 140", frame.getValue(1, 0));

    // existing updated, sorted to bottom (3/40/72)
    ridings.get(5).setResults(Map.of(yp, 280, ndp, 207, lib, 152, grn, 22));
    assertEquals("VUNTUT GWITCHIN", frame.getRowHeader(0));
    assertEquals(Color.RED, frame.getColor(0, 0));
    assertEquals("LIB: 38", frame.getValue(0, 0));
    assertEquals("WATSON LAKE", frame.getRowHeader(1));
    assertEquals(Color.BLUE, frame.getColor(1, 0));
    assertEquals("YP: 150", frame.getValue(1, 0));

    // bottom (out of view) removed (3/40)
    ridings.get(5).setWinner(yp);
    assertEquals("VUNTUT GWITCHIN", frame.getRowHeader(0));
    assertEquals(Color.RED, frame.getColor(0, 0));
    assertEquals("LIB: 38", frame.getValue(0, 0));
    assertEquals("WATSON LAKE", frame.getRowHeader(1));
    assertEquals(Color.BLUE, frame.getColor(1, 0));
    assertEquals("YP: 150", frame.getValue(1, 0));

    // update in view (7/40)
    ridings.get(6).setResults(Map.of(yp, 70, ndp, 3, lib, 77));
    assertEquals("VUNTUT GWITCHIN", frame.getRowHeader(0));
    assertEquals(Color.RED, frame.getColor(0, 0));
    assertEquals("LIB: 77", frame.getValue(0, 0));
    assertEquals("WATSON LAKE", frame.getRowHeader(1));
    assertEquals(Color.BLUE, frame.getColor(1, 0));
    assertEquals("YP: 150", frame.getValue(1, 0));

    // remove from in view (40)
    ridings.get(6).setWinner(lib);
    assertEquals(1, frame.getNumRows());
    assertEquals("WATSON LAKE", frame.getRowHeader(0));
    assertEquals(Color.BLUE, frame.getColor(0, 0));
    assertEquals("YP: 150", frame.getValue(0, 0));

    // add to top (25/40)
    ridings.get(1).setResults(Map.of(yp, 169, ndp, 76, lib, 144));
    assertEquals(2, frame.getNumRows());
    assertEquals("KLUANE", frame.getRowHeader(0));
    assertEquals(Color.BLUE, frame.getColor(0, 0));
    assertEquals("YP: 169", frame.getValue(0, 0));
    assertEquals("WATSON LAKE", frame.getRowHeader(1));
    assertEquals(Color.BLUE, frame.getColor(1, 0));
    assertEquals("YP: 150", frame.getValue(1, 0));

    // update in view, sorted (40/49)
    ridings.get(1).setResults(Map.of(yp, 338, ndp, 153, lib, 289));
    assertEquals(2, frame.getNumRows());
    assertEquals("WATSON LAKE", frame.getRowHeader(0));
    assertEquals(Color.BLUE, frame.getColor(0, 0));
    assertEquals("YP: 150", frame.getValue(0, 0));
    assertEquals("KLUANE", frame.getRowHeader(1));
    assertEquals(Color.BLUE, frame.getColor(1, 0));
    assertEquals("YP: 338", frame.getValue(1, 0));
  }

  private static class Riding extends Bindable {
    enum Property {
      RESULT,
      WINNER
    }

    private final String name;
    private Map<Party, Integer> results = new HashMap<>();
    private Party winner;

    private Riding(String name) {
      this.name = name;
    }

    void setResults(Map<Party, Integer> results) {
      this.results = results;
      onPropertyRefreshed(Property.RESULT);
    }

    void setWinner(Party winner) {
      this.winner = winner;
      onPropertyRefreshed(Property.WINNER);
    }

    Binding<Boolean> isTooClose() {
      return Binding.propertyBinding(
          this,
          t -> t.winner == null && t.results.values().stream().mapToInt(i -> i).sum() > 0,
          Property.RESULT,
          Property.WINNER);
    }

    Binding<Integer> getMargin() {
      return Binding.propertyBinding(
          this,
          t -> {
            List<Integer> topTwoVotes =
                results.values().stream()
                    .sorted(Comparator.reverseOrder())
                    .collect(Collectors.toList());
            if (topTwoVotes.size() == 0) return 0;
            return topTwoVotes.get(0) - topTwoVotes.get(1);
          },
          Property.RESULT);
    }

    Binding<List<Pair<Color, String>>> getBoxes() {
      return Binding.propertyBinding(
          this,
          t ->
              t.results.entrySet().stream()
                  .sorted(Map.Entry.<Party, Integer>comparingByValue().reversed())
                  .map(
                      e ->
                          ImmutablePair.of(
                              e.getKey().getColor(),
                              e.getKey().getAbbreviation() + ": " + e.getValue()))
                  .collect(Collectors.toList()),
          Property.RESULT);
    }
  }
}

File: src/test/java/com/joecollins/graphics/components/MultiSummaryFrameBuilderTest.kt
package com.joecollins.graphics.components

import com.joecollins.bindings.Bindable
import com.joecollins.bindings.Binding
import com.joecollins.bindings.Binding.Companion.fixedBinding
import com.joecollins.bindings.Binding.Companion.propertyBinding
import com.joecollins.graphics.components.MultiSummaryFrameBuilder.Companion.tooClose
import com.joecollins.models.general.Party
import java.awt.Color
import java.util.ArrayList
import java.util.HashMap
import org.junit.Assert
import org.junit.Test

class MultiSummaryFrameBuilderTest {
    @Test
    fun testTooCloseToCallBuilder() {
        val yp = Party("Yukon Party", "YP", Color.BLUE)
        val ndp = Party("New Democratic Party", "NDP", Color.ORANGE)
        val lib = Party("Liberal", "LIB", Color.RED)
        val grn = Party("Green", "GRN", Color.GREEN)
        val ind = Party("Independent", "IND", Color.GRAY)
        val ridings: MutableList<Riding> = ArrayList()
        ridings.add(Riding("Klondike")) // 0
        ridings.add(Riding("Kluane")) // 1
        ridings.add(Riding("Lake Laberge")) // 2
        ridings.add(Riding("Mayo-Tatchun")) // 3
        ridings.add(Riding("Mount Lorne-Southern Lakes")) // 4
        ridings.add(Riding("Pelly-Nisutlin")) // 5
        ridings.add(Riding("Vuntut Gwitchin")) // 6
        ridings.add(Riding("Watson Lake")) // 7
        val frame = tooClose(
                ridings, { it.isTooClose }, { it.margin },
                { fixedBinding(it.name.toUpperCase()) }, { it.boxes },
                2)
                .withHeader(fixedBinding("TOO CLOSE TO CALL"))
                .build()
        Assert.assertEquals("TOO CLOSE TO CALL", frame.header)
        Assert.assertEquals(0, frame.numRows.toLong())

        // add first (36)
        ridings[5].setResults(mapOf(yp to 140, ndp to 104, lib to 76, grn to 11))
        Assert.assertEquals(1, frame.numRows.toLong())
        Assert.assertEquals("PELLY-NISUTLIN", frame.getRowHeader(0))
        Assert.assertEquals(Color.BLUE, frame.getColor(0, 0))
        Assert.assertEquals("YP: 140", frame.getValue(0, 0))

        // add to top (3/36)
        ridings[6].setResults(mapOf(yp to 35, ndp to 2, lib to 38))
        Assert.assertEquals(2, frame.numRows.toLong())
        Assert.assertEquals("VUNTUT GWITCHIN", frame.getRowHeader(0))
        Assert.assertEquals(Color.RED, frame.getColor(0, 0))
        Assert.assertEquals("LIB: 38", frame.getValue(0, 0))
        Assert.assertEquals("PELLY-NISUTLIN", frame.getRowHeader(1))
        Assert.assertEquals(Color.BLUE, frame.getColor(1, 0))
        Assert.assertEquals("YP: 140", frame.getValue(1, 0))

        // add beyond limit (3/36/40)
        ridings[7].setResults(mapOf(yp to 150, ndp to 110, lib to 106, ind to 19))
        Assert.assertEquals(2, frame.numRows.toLong())
        Assert.assertEquals("VUNTUT GWITCHIN", frame.getRowHeader(0))
        Assert.assertEquals(Color.RED, frame.getColor(0, 0))
        Assert.assertEquals("LIB: 38", frame.getValue(0, 0))
        Assert.assertEquals("PELLY-NISUTLIN", frame.getRowHeader(1))
        Assert.assertEquals(Color.BLUE, frame.getColor(1, 0))
        Assert.assertEquals("YP: 140", frame.getValue(1, 0))

        // existing updated, sorted to bottom (3/40/72)
        ridings[5].setResults(mapOf(yp to 280, ndp to 207, lib to 152, grn to 22))
        Assert.assertEquals("VUNTUT GWITCHIN", frame.getRowHeader(0))
        Assert.assertEquals(Color.RED, frame.getColor(0, 0))
        Assert.assertEquals("LIB: 38", frame.getValue(0, 0))
        Assert.assertEquals("WATSON LAKE", frame.getRowHeader(1))
        Assert.assertEquals(Color.BLUE, frame.getColor(1, 0))
        Assert.assertEquals("YP: 150", frame.getValue(1, 0))

        // bottom (out of view) removed (3/40)
        ridings[5].setWinner(yp)
        Assert.assertEquals("VUNTUT GWITCHIN", frame.getRowHeader(0))
        Assert.assertEquals(Color.RED, frame.getColor(0, 0))
        Assert.assertEquals("LIB: 38", frame.getValue(0, 0))
        Assert.assertEquals("WATSON LAKE", frame.getRowHeader(1))
        Assert.assertEquals(Color.BLUE, frame.getColor(1, 0))
        Assert.assertEquals("YP: 150", frame.getValue(1, 0))

        // update in view (7/40)
        ridings[6].setResults(mapOf(yp to 70, ndp to 3, lib to 77))
        Assert.assertEquals("VUNTUT GWITCHIN", frame.getRowHeader(0))
        Assert.assertEquals(Color.RED, frame.getColor(0, 0))
        Assert.assertEquals("LIB: 77", frame.getValue(0, 0))
        Assert.assertEquals("WATSON LAKE", frame.getRowHeader(1))
        Assert.assertEquals(Color.BLUE, frame.getColor(1, 0))
        Assert.assertEquals("YP: 150", frame.getValue(1, 0))

        // remove from in view (40)
        ridings[6].setWinner(lib)
        Assert.assertEquals(1, frame.numRows.toLong())
        Assert.assertEquals("WATSON LAKE", frame.getRowHeader(0))
        Assert.assertEquals(Color.BLUE, frame.getColor(0, 0))
        Assert.assertEquals("YP: 150", frame.getValue(0, 0))

        // add to top (25/40)
        ridings[1].setResults(mapOf(yp to 169, ndp to 76, lib to 144))
        Assert.assertEquals(2, frame.numRows.toLong())
        Assert.assertEquals("KLUANE", frame.getRowHeader(0))
        Assert.assertEquals(Color.BLUE, frame.getColor(0, 0))
        Assert.assertEquals("YP: 169", frame.getValue(0, 0))
        Assert.assertEquals("WATSON LAKE", frame.getRowHeader(1))
        Assert.assertEquals(Color.BLUE, frame.getColor(1, 0))
        Assert.assertEquals("YP: 150", frame.getValue(1, 0))

        // update in view, sorted (40/49)
        ridings[1].setResults(mapOf(yp to 338, ndp to 153, lib to 289))
        Assert.assertEquals(2, frame.numRows.toLong())
        Assert.assertEquals("WATSON LAKE", frame.getRowHeader(0))
        Assert.assertEquals(Color.BLUE, frame.getColor(0, 0))
        Assert.assertEquals("YP: 150", frame.getValue(0, 0))
        Assert.assertEquals("KLUANE", frame.getRowHeader(1))
        Assert.assertEquals(Color.BLUE, frame.getColor(1, 0))
        Assert.assertEquals("YP: 338", frame.getValue(1, 0))
    }

    private class Riding(val name: String) : Bindable<Riding, Riding.Property>() {
        enum class Property {
            RESULT, WINNER
        }

        private var results: Map<Party, Int> = HashMap()
        private var winner: Party? = null

        fun setResults(results: Map<Party, Int>) {
            this.results = results
            onPropertyRefreshed(Property.RESULT)
        }

        fun setWinner(winner: Party?) {
            this.winner = winner
            onPropertyRefreshed(Property.WINNER)
        }

        val isTooClose: Binding<Boolean>
            get() = propertyBinding(
                    this,
                    { it.winner == null && it.results.values.sum() > 0 },
                    Property.RESULT,
                    Property.WINNER)

        val margin: Binding<Int>
            get() = propertyBinding(
                    this,
                    {
                        val topTwoVotes = results.values.sortedDescending()
                        if (topTwoVotes.isEmpty())
                            0
                        else
                            topTwoVotes[0] - topTwoVotes[1]
                    },
                    Property.RESULT)

        val boxes: Binding<List<kotlin.Pair<Color, String>>>
            get() = propertyBinding(
                    this,
                    { me ->
                        me.results.entries.asSequence()
                                .sortedByDescending { it.value }
                                .map {
                                    Pair(
                                            it.key.color,
                                            it.key.abbreviation + ": " + it.value)
                                }
                                .toList()
                    },
                    Property.RESULT)
    }
}
